### PR TITLE
MS Teams Chat With Your Data

### PR DESCRIPTION
**Pull request by t-mbindra**

### CODE REVIEW

File: samples/msteams-chat-with-your-data/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Chat With Your Data
```
Comment (Pawank-MSFT) [inline samples/msteams-chat-with-your-data/README.md:1]: Please refer to the PR below for your reference, including images and sample setup details with format.
https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/1322

**Review by Pawank-MSFT — COMMENTED** [review]
Please follow the reference readme instructions.

File: samples/msteams-chat-with-your-data/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+   ```git clone https://github.com/t-mbindra/chat-with-your-data.git```
+2. Open the project folder(Microsoft-Teams-Samples/samples/msteams-chat-with-your-data) in VS Code.
+3.  Run
+   ```poetry install && poetry build```
+4. Using the Teams Toolkit extension, sign in to your Microsoft 365 account and Azure account under ```ACCOUNTS```.
+5. [Set up your data source using Azure AI resources](#setting-up-your-data-source).
+6. [Populate the environment files](#populating-the-env-files).
+7. Press **Ctrl+Shift+D** to open the ```Run and Debug``` menu. Press ```F5``` or click on the play button.
+
+## Deploying the app on Azure
```
Comment (Pawank-MSFT) [inline samples/msteams-chat-with-your-data/README.md:59]: this step will be in last
https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/1322

**Review by Pawank-MSFT — COMMENTED** [review]
The README requires significant changes. I've included some of these in the PR. Please make sure to follow the guidelines in the README at **"https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/v-pakoshti/ai-meeting-helper/samples/bot-ai-meeting-helper/README.md"** and update your README accordingly.

File: samples/msteams-chat-with-your-data/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+- The user can then interact with the data through the Teams app template.
+  
+## Interacting with the app
+ ![bot-in-action](images/app.png)
+
+## Included Features
+* **Azure Open AI:** Azure OpenAI On Your Data leverages OpenAI's language models, like GPT-3.5 and GPT-4, along with RAG to provide a conversational interface for interacting with your domain data using the chat completion feature.
+* **Azure AI Search:** Utilizes Azure AI Search as a data cource for Azure OpenAI's chat completion feature.
+* **Azure Blob Storage:** The uploaded files are stored in a container in Azure Blob Storage.
+
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md:29]: @Pawank-MSFT The creation is automated via "sh deploy.sh"

File: samples/msteams-chat-with-your-data/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+## Setting up the app in Github Codespaces
+1. Click Open in GitHub Codespaces badge above to create a codespace for the sample app. Wait for the codespace to be setup, it may take a couple of minutes.
+2. Using the Teams Toolkit extension, sign in to your Microsoft 365 account and Azure account under ```ACCOUNTS```.
+3. [Set up your data source using Azure AI resources](#setting-up-your-data-source).
+4. [Populate the environment files](#populating-the-env-files).
+5. Press **Ctrl+Shift+D** to open the ```Run and Debug``` menu. Select ```Debug``` and press ```F5``` or click on the play button.
+6. Download the zip file ```appPackage/build/appPackage.local.zip``` and [sideload the app to Teams](#sideloading-the-app-to-teams).
+
+## Setting up the app locally
+### Prerequisites
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md:39]: @Pawank-MSFT The prerequisites are included here as the other option is "Setting up in GitHub Codespaces", which does not have any prerequisites

File: samples/msteams-chat-with-your-data/README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+This sample showcases a Teams app template to chat with your data via RAG (retrieval augmentation generation) through Azure OpenAI On Your Data and Azure AI Search.
+- The user needs to set up a datasource such as Azure AI Search, Azure Blob Storage, URL/web address, Azure Cosmos DB for MongoDB vCore, uploaded files, or Elasticsearch.
+- The user can then interact with the data through the Teams app template.
+  
+## Interacting with the app
+ ![bot-in-action](images/app.png)
+
+## Included Features
+* **Azure Open AI:** Azure OpenAI On Your Data leverages OpenAI's language models, like GPT-3.5 and GPT-4, along with RAG to provide a conversational interface for interacting with your domain data using the chat completion feature.
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md:26]: Done

File: samples/msteams-chat-with-your-data/README.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+# Chat With Your Data
+[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?hide_repo_select=true&ref=mbindra-ChatWithYourData&repo=322043759&devcontainer_path=.devcontainer%2Fmsteams-chat-with-your-data%2Fdevcontainer.json&resume=1)
+
+This sample showcases a Teams app template to chat with your data via RAG (retrieval augmentation generation) through Azure OpenAI On Your Data and Azure AI Search.
+- The user needs to set up a datasource such as Azure AI Search, Azure Blob Storage, URL/web address, Azure Cosmos DB for MongoDB vCore, uploaded files, or Elasticsearch.
+- The user can then interact with the data through the Teams app template.
+  
+## Interacting with the app
+ ![bot-in-action](images/app.png)
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md]: Done

File: samples/msteams-chat-with-your-data/README.md (hunk 2 above)
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md]: Done

**Review by Pawank-MSFT — COMMENTED** [review]
some small final comments.

File: samples/msteams-chat-with-your-data/README.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+   - Go to the ```teamsazureopenai-search``` resource. Select the ```Keys``` tab under ```Settings```.  Use ```Primary admin key```.
+6. Populate ```AZURE_SEARCH_INDEX``` based on the index name you chose while setting up the data source.
+
+## Sideload the app to Teams
+1. Go to your Teams app and click on the ```Apps``` icon. Select ```Manage your apps``` followed by ```Upload an app```.
+2. Select ```Upload a custom app``` and open the relevant zip file. Click on ```Add``` when prompted.
+
+> If you do not have permission to upload custom apps (sideloading), Teams Toolkit will recommend creating and using a Microsoft 365 Developer Program account - a free program to get your own dev environment sandbox that includes Teams.
+
+## Interact with the app
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md]: Done

File: samples/msteams-chat-with-your-data/README.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+3. Clear the chat history.
+ ![clear-conversation-history](images/clear.png)
+
+## Deploy/Host the app on Azure
+Instead of the ```Debug``` or ```F5``` flow, you can deploy the app on Azure:
+1. [Populate the environment files](#populate-the-environment-files).
+2. Using the Teams Toolkit Extension tab, click on ```Provision``` followed by ```DEPLOY``` under ```LIFECYCLE```. You will be asked to select the subscription and resource group for provisioning.
+3. Download the zip file ```appPackage/build/appPackage.dev.zip``` and [sideload the app to Teams](#sideload-the-app-to-teams).
+
+>[!Note]
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/README.md:104]: Done

**Comment by t-mbindra** [comment]
@microsoft-github-policy-service agree

File: samples/msteams-chat-with-your-data/src/bot.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            default_model=config.AZURE_OPENAI_MODEL,
+            api_version="2024-02-15-preview",
+            endpoint=config.AZURE_OPENAI_ENDPOINT,
+            logger=logger
+        )
+    )
+else: 
+    model = OpenAIModel(
+        AzureOpenAIModelOptions(
+            azure_ad_token_provider=get_bearer_token_provider(DefaultAzureCredential(), 'https://cognitiveservices.azure.com/.default'),
```
[the remark this answers is not in the document]
Comment (t-mbindra) [reply samples/msteams-chat-with-your-data/src/bot.py:52]: It defines the scope for the token request, i.e. what resources the token should have access to

**Review by Pawank-MSFT — APPROVED** [review]
looks correct, Approving!

**Review by Wajeed-msft — APPROVED** [review]

**Comment by Pawank-MSFT** [comment]
Executed build pipeline:
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/10250201673

File: .devcontainer/msteams-chat-with-your-data/devcontainer.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/python
+{
+	"name": "Python 3",
+	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/python:3.11-bullseye",
+
+	// Default workspace folder
+  	"workspaceFolder": "/workspaces/Microsoft-Teams-Samples/samples/msteams-chat-with-your-data/",
```
Comment (N-sources) [inline .devcontainer/msteams-chat-with-your-data/devcontainer.json:9]: wwtg/!5h

File: samples/msteams-chat-with-your-data/.pylintrc [hunk 11]
```diff
@@ -0,0 +1,640 @@
+﻿[MAIN]
+
+# Analyse import fallback blocks. This can be used to support both Python 2 and
+# 3 compatible code, which means that the block might have code that exists
+# only in one or another interpreter, leading to false positives when analysed.
+analyse-fallback-blocks=no
+
+# Clear in-memory caches upon conclusion of linting. Useful if running pylint
```
Comment (N-sources) [inline samples/msteams-chat-with-your-data/.pylintrc:8]: ### 

> [_****_](url)

File: samples/msteams-chat-with-your-data/.pylintrc [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+#errors-only=
+
+# Always return a 0 (non-error) status code, even if lint errors are found.
+# This is primarily useful in continuous integration scripts.
+#exit-zero=
+
+# A comma-separated list of package or module names from where C extensions may
+# be loaded. Extensions are loading into the active Python interpreter and may
+# run arbitrary code.
+extension-pkg-allow-list=
```
Comment (N-sources) [inline samples/msteams-chat-with-your-data/.pylintrc:28]: goo to change the thing !:

**Review by N-sources — COMMENTED** [review]
Moti; creativity/ PRODUCTIVITY//learn..

**Pull request merged**